src/jdk.compiler/share/classes/com/sun/tools/javac/util/Iterators.java

```diff
@@ -43,59 +44,45 @@ public static <I, O> Iterator<O> createCompoundIterator(Iterable<I> inputs, Func
         return new CompoundIterator<>(inputs, converter);
     }
 
-    private static class CompoundIterator<I, O> implements Iterator<O> {
+    private static final class CompoundIterator<I, O> implements Iterator<O> {
 
         private final Iterator<I> inputs;
         private final Function<I, Iterator<O>> converter;
-        @SuppressWarnings("unchecked")
-        private Iterator<O> currentIterator = EMPTY;
+        private Iterator<O> currentIterator = Collections.emptyIterator();
 
         public CompoundIterator(Iterable<I> inputs, Function<I, Iterator<O>> converter) {
             this.inputs = inputs.iterator();
             this.converter = converter;
         }
 
+        @Override
         public boolean hasNext() {
             if (currentIterator != null && !currentIterator.hasNext()) {
                 update();
             }
             return currentIterator != null;
         }
 
+        @Override
         public O next() {
-            if (currentIterator == EMPTY && !hasNext()) {
+            if (!hasNext()) {
                 throw new NoSuchElementException();
             }
             return currentIterator.next();
         }
 
-        public void remove() {
-            throw new UnsupportedOperationException();
-        }
-
         private void update() {
             while (inputs.hasNext()) {
                 currentIterator = converter.apply(inputs.next());
-                if (currentIterator.hasNext()) return;
+                if (currentIterator.hasNext()) return; // implicit null check
             }
             currentIterator = null;
         }
     }
 
-    @SuppressWarnings("rawtypes")
-    private static final Iterator EMPTY = new Iterator() {
-        public boolean hasNext() {
-            return false;
-        }
-
-        @Override
-        public Object next() {
-            return null;
-        }
-    };
-
+    // Iterator is assumed to never return null from next()
```

Comment:
Not sure about this comment - it seems to explain what is wrong in the code that is removed?

Reply:
It's a coincidence: it just happens so that the removed code also implements Iterator.next() by returning null unconditionally. But that comment and the removed code are unrelated.

That comment captures the fact that the filter iterator _below_ that comment relies on the input iterator never to return null from input.next().

Reply:
I clarified that comment in cf90819.